src/daemons/rhsmcertd.c

```diff
@@ -244,34 +247,154 @@ long long gen_random(long long max) {
     return random_num % true_max;
 }
 
-/* Handle program signals */
-void
-signal_handler(int signo) {
-    if (signo == SIGTERM) {
-        /* Close lock file and release lock on this file */
-        if (fd_lock != -1) {
-            close(fd_lock);
-            fd_lock = -1;
+/**
+ * Try to run Python script package-profile-uploader. This script tries to upload DNF profile
+ * to server, when server supports profile. New process is spawned in blocking way.
+ * @return Return true, when uploading was successful. Otherwise, return false.
+ */
+static gboolean
+upload_package_profile ()
+{
+    gboolean ret;
+    const char * argv[] = {PACKAGE_PROFILE_UPLOADER, NULL};
+    gchar *standard_output = NULL;
+    gchar *standard_error = NULL;
+    gint wait_status;
+    GError *error = NULL;
+
+    debug ("Spawning new process of uploading package profile...");
+
+    // Following function will block until process is finished. Other signals
+    // are queued and will be processed, when following process is finished.
+    // It means other signals (SIGTERM) are blocked too. systemd can resolve
+    // this issue. When sending SIGTERM does not terminate daemon, then it
+    // sends SIGKILL signal after some timeout.
+    ret = g_spawn_sync(
+            "/",                // Working directory
+            (gchar**)argv,      // Executable with arguments
+            NULL,               // List of environments is read from parent process
+            G_SPAWN_DEFAULT,    // Default flags
+            NULL,               // No function running before spawning process
+            NULL,               // No argument for such function is needed as well
+            &standard_output,   // stdout if any
+            &standard_error,    // stderr if any
+            &wait_status,       // status if any
+            &error);            // Error if any
+
+    debug ("Uploading of package profile finished");
+
+    if (standard_output != NULL && strlen(standard_output) > 0) {
+        debug ("stdout of uploading package profile: %s", standard_output);
+    }
+    if (standard_error != NULL && strlen(standard_error) > 0) {
+        debug ("stderr of uploading package profile: %s", standard_error);
+    }
+
+    // Error is usually not NULL, when it wasn't possible to span child process
+    // for some reason (e.g. file does not exist)
+    if (error != NULL) {
```

Comment:
here we should directly exit in case `ret` is false, as it means the process failed to run; in case error is not null, printing it as you already did

Reply:
I don't agree. The `ret` is just redundant for `g_spawn_sync()`. It could return void in theory. When you call `g_spawn_sync()` and it failed to run executable, then it return `false` and `error` is not `NULL`.